File: decorator/discount-builder.ts
import { TenOrderDiscount } from "./decorator/ten-order-discount";
import { WelcomeDiscount } from "./decorator/welcome-discount";
import { Discount } from "./discount";
import { DiscountType, STUDENT_DISCOUNT, VIP_DISCOUNT } from "./discount-type";
import { StudentDiscount } from "./student-discount";
import { VipDiscount } from "./vip-discount";

export class DiscountBuilder {
    private discount: Discount

    constructor(discountType: DiscountType) {
        switch(discountType) {
            case STUDENT_DISCOUNT: {
                this.discount = new StudentDiscount();
                return;
            }
            case VIP_DISCOUNT: {
                this.discount = new VipDiscount();
            }
        }
    }

    welcomeDiscount(): DiscountBuilder {
        this.discount = new WelcomeDiscount(this.discount);
        return this;
    }

    tenOrderDiscount(): DiscountBuilder {
        this.discount = new TenOrderDiscount(this.discount);
        return this;
    }

    getDiscount(): Discount {
        return this.discount;
    }
}

File: decorator/discount_builder.go
package decorator

type DiscountBuilder struct {
	discount Discount
}

func NewDiscountBuilder(discountType DiscountType) *DiscountBuilder {
	switch discountType.String() {
	case StudentDiscount.String():
		return &DiscountBuilder{discount: NewStudentDiscount()}
	case VipDiscount.String():
		return &DiscountBuilder{discount: NewVipDiscount()}
	}

	return nil
}

func (d *DiscountBuilder) WelcomeDiscount() *DiscountBuilder {
	d.discount = NewWelcomeDiscount(d.discount)
	return d
}

func (d *DiscountBuilder) TenOrderDiscount() *DiscountBuilder {
	d.discount = NewTenOrderDiscount(d.discount)
	return d
}

func (d DiscountBuilder) GetDiscount() Discount {
	return d.discount
}
